balancer_wrapper.go

```diff
@@ -447,6 +471,37 @@ func (acbw *acBalancerWrapper) RegisterHealthListener(listener func(balancer.Sub
 		if curHD != hd {
 			return
 		}
-		listener(balancer.SubConnState{ConnectivityState: connectivity.Ready})
+		if !healthCheckEnabled {
+			if listener != nil {
+				listener(balancer.SubConnState{ConnectivityState: connectivity.Ready})
+			}
+			return
+		}
+		// Serialize the health updates from the health producer with
+		// other calls into the LB policy.
+		listenerWrapper := func(scs balancer.SubConnState) {
+			acbw.ccb.serializer.TrySchedule(func(ctx context.Context) {
+				if ctx.Err() != nil || acbw.ccb.balancer == nil {
+					return
+				}
+				acbw.healthMu.Lock()
+				curHD := acbw.healthData
+				acbw.healthMu.Unlock()
+				if curHD != hd {
+					return
+				}
+				listener(scs)
+			})
+		}
+
+		if listener == nil {
+			listenerWrapper = nil
+		}
+
+		healthOpts := HealthCheckOptions{
+			HealthServiceName: cfg.ServiceName,
+			Listener:          listenerWrapper,
+		}
+		regHealthLisFn.(func(context.Context, balancer.SubConn, HealthCheckOptions))(ctx, acbw, healthOpts)
```

Comment:
Since `regHealthLisFn` is obtained from the `internal` package, it can only be set by gRPC code. In other places that use vars `internal` with type `any` the code that reads the var is not expecting the assertion to fail.